internal/middleware/owner/owner.go

```diff
@@ -67,3 +72,58 @@ func New(dbs db.Store, usersClient pb.UserServiceClient, logger *zerolog.Logger)
 		return errNotFound
 	}
 }
+
+// AutoPiOwner creates a new middleware handler that checks whether an autopi is paired.
+// For the middleware to allow the request to proceed:
+//
+//   - The request must have a valid JWT, identifying a user.
+//   - There must be a unitID path parameter, and that autopi must exist.
+//   - Either the device has not been paired on chain (anyone can access the endpoint) or
+//     the user has an address on file that is either the owner of the AutoPi or the owner
+//     of the paired vehicle.
+func AutoPi(dbs db.Store, usersClient pb.UserServiceClient, logger *zerolog.Logger) fiber.Handler {
+	return func(c *fiber.Ctx) error {
+		userID := helpers.GetUserID(c)
+		unitID := c.Params("unitID")
+		valid, unitID := services.ValidateAndCleanUUID(unitID)
+		if !valid {
+			return fiber.NewError(fiber.StatusBadRequest, "Unit id is not a valid UUID.")
+		}
+
+		logger := logger.With().Str("userId", userID).Str("unitID", unitID).Logger()
+		c.Locals("userID", userID)
+		c.Locals("unitID", unitID)
+		c.Locals("logger", &logger)
+
+		autopiUnit, err := models.FindAutopiUnit(c.Context(), dbs.DBS().Reader, unitID)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				logger.Info().Msg("AutoPi not minted, or unit ID invalid.")
+				return fiber.NewError(fiber.StatusNotFound, "AutoPi not minted, or unit ID invalid.")
+			}
+			logger.Err(err).Msg("Database failure searching for AutoPi.")
+			return fiber.NewError(fiber.StatusInternalServerError, "Internal error.")
+		}
+
+		// if unit has not been paired or has been unpaired
+		if autopiUnit.PairRequestID.IsZero() || !autopiUnit.UnpairRequestID.IsZero() {
+			logger.Info().Msg("AutoPi is not paired, anyone can access endpoint.")
+			return nil
+		}
+
+		usr, err := usersClient.GetUser(context.Background(), &pb.GetUserRequest{Id: userID})
+		if err != nil {
+			return err
```

Comment:
It would be nice to separate out "user not found" from "5xx error" but it shouldn't happen, this is fine.